test/test_apply_layers.py

```diff
@@ -0,0 +1,273 @@
+
+    layers = [TupleModule().to(self.device) for _ in range(10)]
+    torch_xla.sync()
+
+    scan_layers = deepcopy(layers)
+    loop_layers = deepcopy(layers)
+    torch_xla.sync()
+
+    # Also make input data some non-trivial graph instead of just device data.
+    input_data = (torch.randn(64).to(self.device) * 100,
+                  torch.randn(64).to(self.device) * 200,
+                  torch.randn(64).to(self.device) * 300)
+    a = torch.randn(64).to(self.device)
+    input_data = tuple(t + a for t in input_data)
+    output = apply_layers(scan_layers, input_data)
+    output[0].sum().backward()
+
+    # Test that the result is the same as for loop.
+    loop_output = input_data
+    for layer in loop_layers:
+      loop_output = layer(*loop_output)
+    torch_xla.sync()
+
+    super().compareResults(loop_output, output, abs_err=0.0001, rel_err=0.001)
+    self.assert_different_tensor(loop_output[0], output[0])
+
+    loop_output[0].sum().backward()
+    torch_xla.sync()
+
+    # Test that the gradients are the same too.
+    for layer_scan, layer_loop in zip(scan_layers, loop_layers):
+      assert layer_scan.linear.weight.grad is not None
+      assert layer_loop.linear.weight.grad is not None
+      assert layer_scan.linear.bias.grad is not None
+      assert layer_loop.linear.bias.grad is not None
+      super().compareResults(
+          layer_scan.linear.weight.grad,
+          layer_loop.linear.weight.grad,
+          abs_err=0.0001,
+          rel_err=0.001)
+      super().compareResults(
+          layer_scan.linear.bias.grad,
+          layer_loop.linear.bias.grad,
+          abs_err=0.0001,
+          rel_err=0.001)
+      self.assert_different_tensor(layer_scan.linear.weight.grad,
+                                   layer_loop.linear.weight.grad)
+      self.assert_different_tensor(layer_scan.linear.bias.grad,
+                                   layer_loop.linear.bias.grad)
+
+  def test_decoder_model(self):
+    # Define a decoder model that composes the decoder model in the example,
+    # but adds the ability to run the layers with the `scan` operator.
+    class DecoderOnlyModelWithScan(torch.nn.Module):
+
+      def __init__(self, **kwargs):
+        super(DecoderOnlyModelWithScan, self).__init__()
+        self.decoder = DecoderOnlyModel(**kwargs)
+
+      @property
+      def layers(self) -> Iterable[torch.nn.Module]:
+        return self.decoder.layers
+
+      def forward(
+          self,
+          input_ids: torch.Tensor,
+      ) -> torch.Tensor:
+        return self.decoder.forward(input_ids)
+
+      def forward_scan(
+          self,
+          input_ids: torch.Tensor,
+      ) -> torch.Tensor:
+        inputs_embeds = self.decoder.embed_tokens(input_ids)
+        # embed positions
+        assert isinstance(inputs_embeds, torch.Tensor)
+        # decoder layers
+        hidden_states = apply_layers(self.decoder.layers, inputs_embeds)
+        hidden_states = self.decoder.norm(hidden_states)
```

Comment:
note to myself, I remember we need to add `optimizer_barrier` somewhere in the scan implementation.

Reply:
I believe that's not necessary because scan already strongly constrains the execution ordering. In JAX, I found these comments saying that when checkpoint is used in `scan`, `prevent_cse` is not needed. I can chat offline about this too.